### ISSUE TITLE
Add email verification

### ISSUE THREAD

**Issue by evertiro** [easydigitaldownloads/edd-free-downloads#110]
We've been asked numerous times about ways to secure the download modal. It's about time we do something about it. What are the options that are available? Should we integrate them (any/all) or write free addons that add captcha/whatever to the form? Discuss!

**Comment by pippinsplugins** [comment]
First, email verification and reCaptcha are not the same thing and should be discussed in separate issues. Let's make this issue about email verification.

**Comment by ghost1227** [comment]
That's fair.

**Comment by pippinsplugins** [comment]
For email verification, it's really pretty simple:

1. When modal is submitted, create the payment like we currently do but set the status to `pending`
2. Send an email to the customer with a verification link, just like we do for user verification: https://github.com/easydigitaldownloads/easy-digital-downloads/blob/master/includes/user-functions.php#L655
3. When customer clicks the link emailed to them (confirming their email is real), we update the payment status to `complete`, which then automatically triggers the email receipt.
4. If auto-download is enabled, the file download should start as soon as the verification link processing is completed.

**Comment by dannyb195** [comment]
placing this on hold till #152 gets merged to milestone 2.2.0

**Comment by cklosowski** [comment]
I'm removing 'modal rewrite' since that will not resolve this issue. That label should only be used for items that will be completed within that PR.

**Comment by cklosowski** [comment]
Ok @easydigitaldownloads/core-devs this is ready for some testing. Here's the low-down:

1) Adds a group of settings under the `Processing` section
a) Require verification
b) The message to show in the modal
c) The subject line of the email (supports all EDD Email tags, including the 2 new ones listed next)
d) The email content itself (supports all EDD Email tags, including the 2 new ones listed next)
2) If a user is logged in, email verification is bypassed (we are assuming that a user has a valid email address, however this can always be required by using the filter `edd_free_downloads_require_verification`
3) When the modal is loaded, the message setting is put into an `edd-alert-info` box above the submit button.
4) When the user submits to receive their download, all checks on the required fields are run, and an AJAX request is made to process the free download
a) The free download processing has been updated to check if verification is needed and if it is, the payment is left in a `pending` status until confirmed.
5) Once the processing is done, and we go to send the email
a) If email is successful: The message is changed and the class is altered to `edd-alert-success`
b) On failure: The message is changed and the class is altered to `edd-alert-error`, and the user can try again
6) If successful, an email will be sent to the user (content supplied via the setting above
7) When the link is clicked, validation is run
a) A nonce is present
b) A hash and token are generated to verify that the link clicked belongs to the payment
c) If things are empty, it fails
8) Upon validation:
a) The payment is marked as `publish` (our completed status)
b) The normal file processing is proceeded with, and the On-Complete handler is executed per the setting.

**Comment by mintplugins** [comment]
@cklosowski This is awesome! However, it is giving me a 500 error on the ajax response:

<img width="1440" alt="screen shot 2017-10-26 at 10 55 17 am" src="https://user-images.githubusercontent.com/1638128/32066166-2d5d3ede-ba3c-11e7-98b4-920136d2de73.png">

**Comment by cklosowski** [comment]
@mintplugins can you check your PHP error logs and let me know what the error is please?

**Comment by mintplugins** [comment]
That error was my bad - I had an MU plugin installed that was causing an error. 

This is working great for me. Download link in email works and payment status is set to completed upon verification. 👍

**Comment by pippinsplugins** [comment]
This is all working well for me!

**Comment by pippinsplugins** [comment]
Looks like this isn't quite working in some cases.

I enabled it on restrictcontentpro.com and then tested it. The email is sent fine but I  get an error while verifying the email: `Error processing download of EDD Member Downloads`

URL provided: `https://restrictcontentpro.com/index.php?edd_action=free_download_verify&token=53b900468e8549a5ce7b489492c5f10a&key=c7ba99a4e3b17a147a9d25293ac5dd57&verify_download=bc01f00cab`

I tried with two different email addresses and got the same behavior.

It's getting caught here in `edd_free_download_verify()`:

```php
	if ( empty( $payment_key ) || empty( $token ) || empty( $_GET['verify_download'] ) || ! wp_verify_nonce( $_GET['verify_download'], 'edd_free_downloads_verify' ) ) {
		if ( ! empty( $payment ) ) {
			$download_id   = $payment->downloads[0]['id'];
			$download      = new EDD_Download( $download_id );
			$error_message = sprintf(
				__( 'Error processing download of <a href="%s">%s</a>'),
				get_permalink( $download_id ),
				$download->get_name()
			);
		} else {
			$error_message = __( 'Error processing download. Please contact support.', 'edd-free-downloads' );
		}

		wp_die( $error_message, __( 'Error', 'edd-free-downloads' ), 403 );
	}
```

**Comment by pippinsplugins** [comment]
It's happening on affiliatewp.com too.

**Comment by cklosowski** [comment]
Did you use the same browser you requested the link in? It's a nonced URL.

**Comment by cklosowski** [comment]
If need be, we can remove the nonce.

**Comment by pippinsplugins** [comment]
Oh, no I did not.

Let's remove the nonce. That's going to fail too frequently.

**Comment by cklosowski** [comment]
Ok as long as we know it can lead to verification sharing. Maybe we swap
the nonced for an expiry

**Comment by pippinsplugins** [comment]
Expiry could work well, just like our user verification in core.

**Comment by cklosowski** [comment]
I was thinking of just using our download link expiration time. So the free download verification link only works for whatever your normal download links work for. That way it cannot be shared, and we don't have to introduce a new setting, and it's customizable, and follows expectations of the existing platform.

**Comment by cklosowski** [comment]
Reopening this issue for the tracking of switching to an expiration.

**Comment by mintplugins** [comment]
Tested and this is working as expected. I filled out the modal in incognito, checked and clicked the email verification in non-incognito, and the download worked. Additionally, making any change to the URl at all causes it to fail. 

If I change the token, it gives a general error
`Error processing download. Please contact support.`

If I change the ttl it gives a product-specific error:
`Error processing download of Product F`

**Comment by cklosowski** [comment]
Closing this out as it was merged an in beta, and seems to be working well.
